### PR TITLE
Add dictionary unions (| and |=)

### LINKED ISSUE #14: support dict concatenation: dict | dict
Bazel has been supporting dictionary concatenation with the `+` operator for a long time. Python doesn't.

We've deprecated the operator for a long time, mostly because of issues in Skydoc. But the original Skydoc is going away, and it's unclear to me if we should proceed with the removal of the operator (as we rely less and less on Python).

https://github.com/bazelbuild/bazel/issues/6461

Any opinion?

### PR DESCRIPTION
**Pull request by fmeum**
Closes #14

### CODE REVIEW

**Comment by fmeum** [comment]
I submitted a Java implementation, including test cases, as https://github.com/bazelbuild/bazel/pull/14540.

File: spec.md [hunk 1]
```diff
@@ -1027,12 +1027,20 @@ fail.
...
-`dict.update` method instead.
+The binary `|` operation may be applied to two dictionaries. It yields a new
+dictionary whose set of keys is the union of the sets of keys of the two
+operands. The corresponding values are taken from the operands, where the value
+taken from the right operand takes precedence if both contain a given key.
+Iterating over the keys in the resulting dictionary first yields all keys in
+the left operand in insertion order, then all keys in the right operand that
+were not present in the left operand, again in insertion order.
+<b>Note:</b> Previously, the binary `+` operator was used for dictionary unions.
+This is now deprecated.
```
Comment (laurentlb) [inline spec.md:1038]: Remove the "note": it was not part of the spec

**Review by laurentlb — APPROVED** [review]
lgtm

File: spec.md (hunk 1 above)
Comment (fmeum) [reply spec.md]: Dropped it.

**Comment by laurentlb** [comment]
cc @brandjon @adonovan

**Comment by stepancheg** [comment]
Can we undo this change before it is too late?

* syntax contradicts meaning of a union: result is not a union of two dicts if keys intersect
* python does not support it, but support `{**dict1, **dict2}` which is unambiguous

**Comment by fmeum** [comment]
> Can we undo this change before it is too late?
> 
> * syntax contradicts meaning of a union: result is not a union of two dicts if keys intersect
> * python does not support it, but support `{**dict1, **dict2}` which is unambiguous

This operator, with the same name and semantics, has been part of Python since [PEP-584](https://www.python.org/dev/peps/pep-0584), which also has an extensive section weighing the pros and cons. I am also not fully happy with the name "union", given that dictionaries are naturally maps of sets rather than sets, but Python provides the precedent.

**Comment by stepancheg** [comment]
> the name "union"

If it is called “or” (which is another form of union) it is equally inconsistent.

> has been part of Python since [PEP-584](https://www.python.org/dev/peps/pep-0584)

TIL. I would not do it, but if Python has done it, it is OK.

Thank you for the explanation!

**Comment by tkoeppe** [comment]
This feature is very useful for combining `select()`-ed pieces of a dictionary with unconditional ones.

**Pull request merged**